demo.py

```diff
@@ -0,0 +1,146 @@
+class Resource(models.Model):
+    client = ForeignKey(Client)
+    owned_by = ForgeignKey(Resource, null=True)
+    name = CharField(blank=False, max_length=128)
+    configuration = JSONField(null=False)
+    types = ResourceTypeRegistry()
+    owned_resources = ResourceManger()
+
+    @property
+    def configuration_schema(self):
+        """ Define the options used to configure this resource. """
+        return {}  # No configurable options by default
+
+    @property
+    def state(self):
+        """
+        The current state of this resource.
+        The return value is always a subclass of ResourceState, and typically will be a subclass of:
+          NeedsRefreshState, ApplyingConfigurationState, NeedsInputState, ReadyState, ErrorState, or DestroyingState
```

Comment:
States are already more fine-grained than this - cf https://github.com/open-craft/opencraft/blob/545e0a2d43358f9ca0815e0aa9c19c62a5a144ce/instance/models/server.py#L99-L122

Also, there are two different notions in terms of status - the step it is in ("provisioning", "booting", etc.), and its progression/result within that step (running, completed, error...). Using classes could be a good way to refine or combine these - but with the list above it's not clear how we would express things like "Error while applying configuration" (which would be a combination of ApplyingConfigurationState and ErrorState, and is different from "Error while destroying" or "Error while rebooting" for example). Or "Ready but needs refresh" vs "Error but needs refresh". It's good to refactor, but it would be good to start from the existing implementation, to focus on changing the parts for which we have identified a specific improvement need.